### DISCUSSION TITLE
Not sure Sparkle being initialized

### DISCUSSION THREAD

**Discussion by coryl** [sparkle-project/Sparkle#2486] [Q&A] [answered]
### Description of the problem
I'm following the [Sparkle setup documentation](https://github.com/sparkle-project/Sparkle#troubleshooting) by creating an Object in Storyboard, and wiring a `NSMenuItem` to the `checkForUpdates` action.

I'm at the phase now where I'm testing. I've lowered my build version, published my appcast.xml, uploaded my .zip, and I'm ready to go. Clicking "Check for Updates" works perfectly.

However I'm trying to understand Sparkle's default behaviour (I want to see the automatic ask for permission after loading the app twice like the documentation says), however I can't trigger it.

To do that, I'm wiping my all my SU_____ keys in my preferences plist (`/Users/user/Library/Preferences/com.myorg.myapp.plist`). Then I'm starting the app (in Xcode), quitting the app, again and again. I noticed no new entries into my preferences plist ever gets added unless I manually run `checkForUpdates` via the menu item.

My app's Info.plist has no keys set for now, as I just want to see the default behaviour.

Its almost as if Sparkle is never running, never gets initialized. I want to see Sparkle check for updates on its own, or at least ask me to check automatically. What am I missing here?

### Do you use Sandboxing in your app?
No

### Version of `Sparkle.framework` in the latest version of your app
2.5.2

### Version of `Sparkle.framework` in the old version of app that your users have (or N/A)
N/A

### Sparkle's output from Console.app
Not sure if this is relevant, but I see for process: kernel
```
Library Validation failed: Rejecting '/Users/user/Library/Developer/Xcode/DerivedData/myapp-cgooyrpjralhmpbjlpowzoomzkoa/Build/Products/Debug/Sparkle.framework/Versions/B/Sparkle' (Team ID: none, platform: no) for process 'myapp(97721)' (Team ID: EUFMJ95U79, platform: no), reason: mapping process and mapped file (non-platform) have different Team IDs
```

**Comment by zorgiepoo** [comment, marked as answer]
> To do that, I'm wiping my all my SU_____ keys in my preferences plist (/Users/user/Library/Preferences/com.myorg.myapp.plist). Then I'm starting the app (in Xcode), quitting the app, again and again. I noticed no new entries into my preferences plist ever gets added unless I manually run checkForUpdates via the menu item.


You can't manipulate the defaults on disk like that. You need to use the `defaults` command to remove defaults. They are otherwise persisted in-memory in a daemon. If you `defaults delete my-bundle-id` then on 2nd launch Sparkle should ask the permission prompt for checking for updates automatically. Some other details are in the testing section on the main website documentation page. There may also be defaults issues if you had a sandboxed app and later make it a non-sandboxed..

> Library Validation failed: Rejecting '/Users/user/Library/Developer/Xcode/DerivedData/myapp-cgooyrpjralhmpbjlpowzoomzkoa/Build/Products/Debug/Sparkle.framework/Versions/B/Sparkle' (Team ID: none, platform: no) for process 'myapp(97721)' (Team ID: EUFMJ95U79, platform: no), reason: mapping process and mapped file (non-platform) have different Team IDs

Our website documentation explains this:

> If you enable Library Validation, which is part of the Hardened Runtime and required for notarization, you will also need to either sign your application with an Apple Development certificate for development (requires being in Apple’s developer program), or disable library validation for Debug configurations only. Otherwise, the system may not let your application load Sparkle if you attempt to sign to run locally via an ad-hoc signature. This is not an issue for distribution when you sign your application with a Developer ID certificate.

You should have Sparkle configured to Code-sign on Copy. Interesting the signatures aren't matching in your error.

**Reply by coryl** [reply]
Thanks for your reply. I used the delete command like you said (`defaults delete my-bundle-id`) and I was able to successfully trigger the default permissions pop up simply by running the app:

![image](https://github.com/sparkle-project/Sparkle/assets/1355864/d8635fe9-2119-4d2c-a3e5-89f05b3d3368)

However, my problem now is that clicking on either of those buttons does not do anything, the box simply remains and does not dismiss or take any action.

I began to look into the library validation issues like you mentioned. I am using Xcode's Swift Package Manager to use Sparkle.

I thought I was signing with my development certificate, here's what I currently have:
![image](https://github.com/sparkle-project/Sparkle/assets/1355864/836f409b-c823-4677-acab-6376a71a2fb3)
![image](https://github.com/sparkle-project/Sparkle/assets/1355864/9dab817b-1a1e-4a21-9c67-e94a143125a4)

But regardless, I first tried creating a debug only entitlements file and set in Build Settings > Entitlements > Debug:
```
<dict>
	<key>com.apple.security.cs.disable-library-validation</key>
	<true/>
    <key>com.apple.security.app-sandbox</key>
    <false/>
    <key>com.apple.security.files.user-selected.read-write</key>
    <true/>
</dict>
```
That didn't seem to do anything, button clicks still don't do anything, and still seeing library validation error in console.app

So I just tried manually forcing the disable library validation:
![image](https://github.com/sparkle-project/Sparkle/assets/1355864/2d357cd4-9caf-4b72-8631-ea01c8a57754)

This time the buttons still fail to do anything, but I'm not seeing library validation errors, but am I seeing some bundle (not loaded) messages:
`Bundle: CFBundle 0x600003cff560 </Users/user/Library/Developer/Xcode/DerivedData/frankenfile-cgooyrpjralhmpbjlpowzoomzkoa/Build/Products/Debug/Sparkle.framework> (not loaded), key: AUc-33-qGN.title, value: Automatically download and install updates, table: SUUpdatePermissionPrompt, localizationNames: (null), result: Automatically download and install updates`

Any other things I can try? Thank you

**Reply by coryl** [reply]
Also, not sure if its relevant, but if I go through to click my menu bar item's "Check For Updates", I am able to get a dialog box that does work (able to download and update, able to skip). However the previous initial box still lingers and takes no action:
![image](https://github.com/sparkle-project/Sparkle/assets/1355864/46fec221-90f1-4501-9a16-f51d0c87ed9a)

I'm able to execute and receive these update dialog boxes regardless of whether I enabled/disabled library validation.

**Reply by zorgiepoo** [reply]
How are you instantiating  the updater? You should be instantiating the updater once and keeping a reference around. It should not be possible for those two dialogs to appear simultaneously unless they are different updater instances. 

Be mindful that there is no singleton accessor.

**Reply by coryl** [reply]
You're right, I found some old debugging code I left in the app delegate. After removing that, I'm able to click the default permission prompt buttons. Also don't seem to need any library validation stuff anymore, not getting any console errors.

So I'm back to default behaviour for the framework. Thanks!

**Comment by coryl** [comment]
> > To do that, I'm wiping my all my SU_____ keys in my preferences plist (/Users/user/Library/Preferences/com.myorg.myapp.plist). Then I'm starting the app (in Xcode), quitting the app, again and again. I noticed no new entries into my preferences plist ever gets added unless I manually run checkForUpdates via the menu item.
> 
> You can't manipulate the defaults on disk like that. You need to use the `defaults` command to remove defaults. They are otherwise persisted in-memory in a daemon. If you `defaults delete my-bundle-id` then on 2nd launch Sparkle should ask the permission prompt for checking for updates automatically. Some other details are in the testing section on the main website documentation page. There may also be defaults issues if you had a sandboxed app and later make it a non-sandboxed..
> 
> > Library Validation failed: Rejecting '/Users/user/Library/Developer/Xcode/DerivedData/myapp-cgooyrpjralhmpbjlpowzoomzkoa/Build/Products/Debug/Sparkle.framework/Versions/B/Sparkle' (Team ID: none, platform: no) for process 'myapp(97721)' (Team ID: EUFMJ95U79, platform: no), reason: mapping process and mapped file (non-platform) have different Team IDs
> 
> Our website documentation explains this:
> 
> > If you enable Library Validation, which is part of the Hardened Runtime and required for notarization, you will also need to either sign your application with an Apple Development certificate for development (requires being in Apple’s developer program), or disable library validation for Debug configurations only. Otherwise, the system may not let your application load Sparkle if you attempt to sign to run locally via an ad-hoc signature. This is not an issue for distribution when you sign your application with a Developer ID certificate.
> 
> You should have Sparkle configured to Code-sign on Copy. Interesting the signatures aren't matching in your error.
